internal/session/user_session.go

```diff
@@ -12,3 +15,50 @@ func NewDefaultUserSession() UserSession {
 		LastActivity:        0,
 	}
 }
+
+// SetOneFactor sets the expected property values for one factor authentication.
+func (s *UserSession) SetOneFactor(now time.Time, details *authentication.UserDetails, keepMeLoggedIn bool) {
+	if s.FirstFactorAuthn == 0 {
+		s.FirstFactorAuthn = now.Unix()
+	}
+
+	s.LastActivity = now.Unix()
+	s.AuthenticationLevel = authentication.OneFactor
+
+	s.KeepMeLoggedIn = keepMeLoggedIn
+
+	s.Username = details.Username
+	s.DisplayName = details.DisplayName
+	s.Groups = details.Groups
+	s.Emails = details.Emails
+}
+
+// SetTwoFactor sets the expected property values for two factor authentication.
+func (s *UserSession) SetTwoFactor(now time.Time) {
+	if s.SecondFactorAuthn == 0 {
+		s.SecondFactorAuthn = now.Unix()
+	}
+
+	s.LastActivity = now.Unix()
+	s.AuthenticationLevel = authentication.TwoFactor
+}
+
+// AuthenticatedAt returns the unix timestamp this session authenticated successfully at the given level.
+func (s UserSession) AuthenticatedAt(level authorization.Level) (authenticatedAt time.Time) {
+	switch level {
+	case authorization.OneFactor:
+		return time.Unix(s.FirstFactorAuthn, 0)
+	case authorization.TwoFactor:
+		return time.Unix(s.SecondFactorAuthn, 0)
+	}
+
+	if s.SecondFactorAuthn != 0 {
```

Comment:
In my opinion and unless I missed something, this function should raise an error if we are not matching a known state (what is covered by the switch block). It's better to advertise the admin when something is not happening as expected but for sure we should not let erroneous cases execute silently with best-effort values. It might be ok in a non-security oriented component but here in my opinion we should be on the safe side if that makes sense. What do you think?